src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1837,17 +1838,20 @@ RemoveCoordinatorPlacement(List *placementList)
  */
 static List *
 SingleShardSelectTaskList(Query *query, uint64 jobId, List *relationShardList,
-						  List *placementList,
-						  uint64 shardId)
+						  List *placementList, uint64 shardId)
 {
 	Task *task = CreateTask(SELECT_TASK);
-	StringInfo queryString = makeStringInfo();
 	List *relationRowLockList = NIL;
 
 	RowLocksOnRelations((Node *) query, &relationRowLockList);
-	pg_get_query_def(query, queryString);
 
-	task->queryString = queryString->data;
+	/*
+	 * For performance reasons, we skip generating the queryString. For local
+	 * execution this is not needed, so we wait until the executor determines
+	 * that the query cannot be executed locally.
+	 */
+	task->query = query;
```

Comment:
I think the  down side of doing this is that now the `task` is heavier, in the sense that when we need to copy the task, we'd be copying  the queryTree, instead of a string. (Similar for `SingleShardModifyTaskList `). So, is it worth *always* doing it for single shard queries?

I don't want to over complicate things, just sharing some potential ways of doing it less commonly.

So, is it possible that we only do this when there is a possibility of local execution? Something like:


```C

/* best effort to defer query string generation */
if (TaskAccessesLocalNode())
{
task->query = query;
task->queryStringLazy = NULL;
}
else
{
pg_get_query_def(query, queryString);
task->queryString = queryString->data;
}
```

We're not doing deferring query generation for multi-shard queries anyway. So, I wouldn't mind doing it only for `tasks that could potentially use local execution`

What is your take on this?


(There is already a function named `TaskAccessesLocalNode` for checking this, we could change the api to get a placement list instead of task)

Reply:
Last commit implements this in a helper function.